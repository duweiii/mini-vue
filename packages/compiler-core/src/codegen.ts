import { ENodeType } from "./ast";
import { CREATE_ELEMENT_VNODE, CREATE_TEXT_VNODE, helperMap, TO_DISPLAY_STRING } from "./runtimeHelpers";

export function generate(node){
  const context = createCodegenContext()
  const { push } = context;
  const functionName = 'render';
  const args = ['_ctx', '_cache'];
  const signature = args.join(',');
  
  if( node.helpers.length > 0 ){
    genFunctionPreamble(node, context)
  }

  push('return function ')
  push(`${functionName}(${signature}){return `)
  genNode(node.codegenNode, context)
  push("}")

  return {
    code: context.code
  }
}

function genFunctionPreamble(node, context){
  const { push } = context;
  const VueBinging = 'Vue';
  const aliasHelpers = (s) => `${helperMap[s]}:_${helperMap[s]}`;
  const helpers = node.helpers;
  const declaration = `const { ${helpers.map(aliasHelpers)} } = ${VueBinging}`
  push(declaration)
  push('\n')
}

function createCodegenContext(){
  const context = {
    code: '',
    push(source){
      context.code += source
    },
    helper(key){
      return `_${helperMap[key]}`;
    }
  }
  return context;
}

function genNode(node, context){
  switch( node.type ){
    case ENodeType.TEXT:
      genText(node, context);
      break;
    case ENodeType.INTERPOLATION:
      genInterpolation(node, context)
      break;
    case ENodeType.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break;
    case ENodeType.ELEMENT:
      genElement(node, context);
      break;
    case ENodeType.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;
    default:
      break;
  }
}

function genCompoundExpression(node, context){
  textNodeWrapper(node, context, () => {
    const { children } = node;
    const { push } = context;
    for (let i = 0; i < children.length; i++) {
      let child = children[i];
      if( typeof child === 'string' ){
        push(child);
      }else{
        genNode(child, context)
      }
    }
  })
}

function genElement(node, context){
  const { helper, push } = context;
  const { tag, props } = node;
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  genNodeList( genNullable([tag, props]), context );
  genElementChildren(node, context);
  push(")")
}

function genElementChildren(node, context){
  const { children } = node;
  const { push } = context;
  if( children.length !== 0 ){
    push(', ')
  }
  const { addArrayWrapper } = node;
  addArrayWrapper && push('[');
  const length = children.length;
  for (let i = 0; i < length; i++) {
    const child = children[i];
    genNode(child, context);
    if( i < length -1 ){
      push(', ')
    }
  }
  addArrayWrapper && push(']');
}
function genNullable(args){
  return args.map(arg => arg || 'null');
}

function genNodeList(nodes, context){
  const { push } = context;
  for (let index = 0; index < nodes.length; index++) {
    const node: any = nodes[index];
    push(node);
    if( index < nodes.length - 1 ){
      push(', ')
    }
  }
}

function genText(node, context){
  textNodeWrapper(node, context, () => {
    const { push } = context;
    push(`'${node.content}'`);
  } )
}

function genInterpolation(node, context){
  textNodeWrapper(node, context, () => {
    const { push, helper } = context;
    push(`${helper(TO_DISPLAY_STRING)}(`);
    genNode( node.content, context )
    push(')')
  })
}

function genExpression(node, context){
  const { push } = context;
  push( node.content )
}

function textNodeWrapper(node, context, fn){
  const { addTextWrapper } = node;
  const { push, helper } = context;
  addTextWrapper && push(`${helper(CREATE_TEXT_VNODE)}(`)
  fn();
  addTextWrapper && push(')')
}