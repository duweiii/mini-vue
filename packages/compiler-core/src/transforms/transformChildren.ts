import { CREATE_TEXT_VNODE } from "../runtimeHelpers";
import { isElement, shouldWrapTextNode } from "../utils";

export function transformChildren(node, context){
  return () => {
    if( isElement(node) ){
      const { children } = node;
      const length = children.length;
      if( length === 1 ){
        /**
         * 如果只有一个child，但是child类型为element，也是需要放到数组中处理的。
         */
        const child = children[0];
        if( isElement(child) ){
          node.addArraySymbol = true;
        }
      }else if ( length > 1 ){
        /**
         * 遍历children，对其中的 text/interpolation/compound,需要用 createTextVNode 包裹
         * 仅对 elementNode.children 中的文本节点添加 createTextVNode 包裹的标记
         * 
         * 因为在runtime-core中，处理children只识别string/array类型，在这里，只有一个string是可以渲染的。
         * 在array类型的children中，遍历到每一个child，去patch，在patch中，只有一个string，就无法正常渲染了，
         * 所以需要通过createTextVNode包裹。
         */
        node.addArrayWrapper = true;
        for (let i = 0; i < children.length; i++) {
          const child = children[i];
          if( shouldWrapTextNode(child) ){
            child.addTextWrapper = true;
            context.helper(CREATE_TEXT_VNODE)
          }
        }
      }
    }
  }
}