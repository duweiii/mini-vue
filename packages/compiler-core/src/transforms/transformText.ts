import { ENodeType } from "../ast";
import { isElement, isText } from "../utils";

export function transformText(node){
  if( isElement(node) ){
    return () => {
      const { children } = node;
      let currentContainer;

      for( let i = 0; i < children.length; i++ ){
        let child = children[i]
        if( isText(child) ){
          for( let j = i + 1; j < children.length; j++ ){
            let next = children[j];
            if( isText(next) ){
              if( !currentContainer ){
                currentContainer = children[i] = {
                  type: ENodeType.COMPOUND_EXPRESSION,
                  children: [child]
                }
              }

              currentContainer.children.push(' + ');
              currentContainer.children.push( next );
              children.splice(j, 1);
              j--;
            }else{
              currentContainer = undefined;
              break;
            }
          }
        }
      }
    }
  }
}