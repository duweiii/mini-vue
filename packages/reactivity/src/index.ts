export { reactive, readOnly, shallowReadonly, isReactive, isProxy, isReadonly } from './reactive'
export { ref, isRef, unRef, proxyRefs } from './ref'
export { effect } from './effect';
export { watch } from './watch'