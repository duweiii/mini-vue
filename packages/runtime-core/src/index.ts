export { h } from "./h";
export { renderSlots } from './helpers/renderSlots'
export { createTextVNode, createElementVNode } from './createVNode'
export { getCurrentInstance, registerCompiler } from './component'
export { provide, inject } from './apiInject'
export { createRenderer } from './render'
export { nextTick } from './schedulers'
export { toDisplayString } from '@guide-mini-vue/shared';
export * from '@guide-mini-vue/reactivity';
export * from '@guide-mini-vue/shared';
