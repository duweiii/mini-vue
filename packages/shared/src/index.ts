export * from './toDisplayString';
export * from './shapeFlags';
export const extend = Object.assign;

export const isObject = (value) => {
  return typeof value === 'object';
}
export const isString = (value) => {
  return typeof value === 'string';
}
export const EMPTY_OBJECT = {};
export const hasChanged = (value, newValue) => {
  return !Object.is(value, newValue);
}

export const hasOwn = (object: {[key: string]: any}, key: string) => {
  return Object.prototype.hasOwnProperty.call(object, key);
}

export const camelize = (str: string) => {
  return str.replace(/-(\w)/g, (_, c: string ) => {
    return c ? c.toUpperCase() : '';
  })
}
export const capitalize = (str: string): string => {
  return str.charAt(0).toUpperCase() + str.slice(1);
}
export const toHandleKey = (str: string): string => {
  return 'on' + capitalize(str);
}