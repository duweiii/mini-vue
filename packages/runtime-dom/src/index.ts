export * from '@guide-mini-vue/runtime-core';
import { createRenderer } from "@guide-mini-vue/runtime-core";

function createElement(type: string){
  return document.createElement(type);
}
function patchProp(el, key, prevValue, nextValue){
  const isOn = (key: string) => /^on[A-Z]/.test(key);
  if( isOn(key) ){
    let eventType = key.slice(2).toLowerCase();
    el.addEventListener(eventType, nextValue);
  } else {
    if( nextValue === undefined || nextValue === null ){
      el.removeAttribute(key)
    }else{
      el.setAttribute(key, nextValue)
    }
  }
}
function insert(el, parent, anchor = null){
  return parent.insertBefore(el, anchor)
}
function setElementText(el, text){
  el.textContent = text;
}
function remove(el){
  const parentNode = el.parentNode;
  if( parentNode ){
    parentNode.removeChild(el)
  }
}

const renderer: any = createRenderer({
  createElement,
  patchProp,
  insert,
  setElementText,
  remove
})

export function createApp(...args){
  return renderer.createApp(...args)
}