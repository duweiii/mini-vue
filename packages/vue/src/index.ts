export * from '@guide-mini-vue/runtime-dom'
import { baseCompile } from '@guide-mini-vue/compiler-core';
import * as runtimeDom from '@guide-mini-vue/runtime-dom';
import { registerCompiler } from '@guide-mini-vue/runtime-dom';

function compileToFunction(template: string){
  const { code } = baseCompile(template);

  const render = new Function('Vue', code)(runtimeDom);

  return render;
}

registerCompiler(compileToFunction);