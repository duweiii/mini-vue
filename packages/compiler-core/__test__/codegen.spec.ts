import { generate } from "../src/codegen";
import { baseParse } from "../src/parse"
import { transform } from "../src/transform";
import { transformChildren } from "../src/transforms/transformChildren";
import { transformElement } from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformText } from "../src/transforms/transformText";

describe('code generate', () => {
  it('happy path: text node', () => {
    const ast = baseParse('hi');
    transform(ast)
    const { code } = generate( ast );
    expect( code ).toMatchSnapshot();
  })

  it('happy path: interpolation node', () => {
    const ast = baseParse('{{message}}');
    transform(ast, { nodeTransforms: [transformExpression]})
    const { code } = generate( ast );
    expect( code ).toMatchSnapshot();
  })

  it('happy pathï¼šelement node', () => {
    const ast = baseParse('<div>ni</div>');
    transform(ast, { nodeTransforms: [transformElement]})
    const { code } = generate( ast );
    expect( code ).toMatchSnapshot();
  })

  it("happy path: unit", () => {
    const ast = baseParse("<div>hi,{{message}}</div>")
    transform(ast, { nodeTransforms: [ transformExpression, transformElement, transformText]})

    const { code } = generate( ast );
    expect( code ).toMatchSnapshot();
  })
  
  it("happy path: multiple element", () => {
    const ast = baseParse("<div>hi,{{message}}<p>embed-element</p></div>")
    transform(ast, { nodeTransforms: [ transformExpression, transformChildren, transformElement, transformText]})
    const { code } = generate( ast );
    expect( code ).toMatchSnapshot();
  })

})