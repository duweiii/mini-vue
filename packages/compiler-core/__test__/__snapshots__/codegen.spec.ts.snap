// Vitest Snapshot v1

exports[`code generate > happy path: interpolation node 1`] = `
"const { toDisplayString:_toDisplayString } = Vue
return function render(_ctx,_cache){return _toDisplayString(_ctx.message)}"
`;

exports[`code generate > happy path: multiple element 1`] = `
"const { toDisplayString:_toDisplayString,createElementVNode:_createElementVNode,createTextVNode:_createTextVNode } = Vue
return function render(_ctx,_cache){return _createElementVNode('div', null, [_createTextVNode('hi,' + _toDisplayString(_ctx.message)), _createElementVNode('p', null, 'embed-element')])}"
`;

exports[`code generate > happy path: text node 1`] = `"return function render(_ctx,_cache){return 'hi'}"`;

exports[`code generate > happy path: unit 1`] = `
"const { toDisplayString:_toDisplayString,createElementVNode:_createElementVNode } = Vue
return function render(_ctx,_cache){return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`code generate > happy path：element node 1`] = `
"const { createElementVNode:_createElementVNode } = Vue
return function render(_ctx,_cache){return _createElementVNode('div', null, 'ni')}"
`;

exports[`code generate happy path: interpolation node 1`] = `
"const { toDisplayString:_toDisplayString } = Vue
return function render(_ctx,_cache){return _toDisplayString(_ctx.message)}"
`;

exports[`code generate happy path: multiple element 1`] = `
"const { toDisplayString:_toDisplayString,createElementVNode:_createElementVNode,createTextVNode:_createTextVNode } = Vue
return function render(_ctx,_cache){return _createElementVNode('div', null, [_createTextVNode('hi,' + _toDisplayString(_ctx.message)), _createElementVNode('p', null, 'embed-element')])}"
`;

exports[`code generate happy path: text node 1`] = `"return function render(_ctx,_cache){return 'hi'}"`;

exports[`code generate happy path: unit 1`] = `
"const { toDisplayString:_toDisplayString,createElementVNode:_createElementVNode } = Vue
return function render(_ctx,_cache){return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`code generate happy path：element node 1`] = `
"const { createElementVNode:_createElementVNode } = Vue
return function render(_ctx,_cache){return _createElementVNode('div', null, 'ni')}"
`;
